roles/openshift_health_checker/library/search_journalctl.py

```diff
@@ -0,0 +1,163 @@
+    """Receives iterable input, a matcher dictionary and scans journalctl at the
+    given systemd module for log messages matching a given regular expression.
+    Returns the matched log message (or None) and an error string if any."""
+    try:
+        regexp = re.compile(matcher["regexp"])
+        start_regexp = re.compile(matcher["start_regexp"])
+    except re.error as err:
+        msg = "A log matcher object was provided with an invalid regular expression: {}"
+        raise InvalidMatcherRegexpException(msg.format(str(err)))
+
+    log_count = 0
+    matched = None
+
+    for line in log_input:
+        log_count += 1
+        if log_count >= log_count_limit:
+            break
+
+        try:
+            obj = json.loads(line.rstrip())
+
+            # don't need to look past the most recent service restart
+            if start_regexp.match(obj["MESSAGE"]):
+                break
+
+            if stamp_too_old(obj["__REALTIME_TIMESTAMP"], timestamp_limit_seconds):
+                break
+
+            if regexp.match(obj["MESSAGE"]):
+                matched = line.rstrip()
+                break
+
+        except ValueError as err:
+            return None, str(err)
+
+    return matched, None
+
+
+def ensure_matcher_fields(matcher):
+    """Ensures that expected fields exist on a user-defined matcher dict"""
+    required_fields = [
+        "start_regexp",
+        "regexp",
+        "unit",
+    ]
+    missing_fields = set(required_fields) - set([field for field in matcher])
+    if len(missing_fields):
+        raise InvalidMatcherException("Matcher missing required fields", list(missing_fields))
+
+
+def _get_log_matches(matchers, log_count_limit, timestamp_limit_seconds):
+    matched_regexp = []
+    errors = []
+
+    for matcher in matchers:
+        try:
+            ensure_matcher_fields(matcher)
+        except InvalidMatcherException as err:
+            errors.append(str(err))
+            continue
+
+        log_input, error = _get_log_input(matcher)
+        if error:
+            errors.append(error)
+            continue
+
+        try:
+            matched, error = find_matches(log_input, matcher, log_count_limit, timestamp_limit_seconds)
+            if matched:
+                matched_regexp.append(matcher["regexp"])
+                continue
+        except InvalidMatcherRegexpException as err:
+            error = str(err)
+
+        if error:
+            errors.append(error)
+
+    return matched_regexp, errors
+
+
+def _get_log_input(matcher):
+    try:
+        cmd_output = subprocess.Popen(list([
+            '/bin/journalctl',
+            '-ru', matcher["unit"],
+            '--output', 'json',
+        ]), stdout=subprocess.PIPE)
+
+        return iter(cmd_output.stdout.readline, ''), None
+
+    except subprocess.CalledProcessError as exc:
+        err = '[rc {}] {}\n{}'.format(exc.returncode, ' '.join(exc.cmd), exc.output)
```

Comment:
yes, this error is propagated up to `_get_log_matches`, and returned in the overall list of errors. Will improve message